oauthproxy_test.go

```diff
@@ -1786,6 +1786,124 @@ func TestGetRedirect(t *testing.T) {
 			assert.Equal(t, tt.expectedRedirect, redirect)
 		})
 	}
+
+	proxyOpts := baseTestOptions()
+	proxyOpts.ReverseProxy = true
+	proxyOpts.WhitelistDomains = append(proxyOpts.WhitelistDomains, ".example.com")
+	proxyOpts.WhitelistDomains = append(proxyOpts.WhitelistDomains, ".example.com:8443")
+	err = validation.Validate(proxyOpts)
+	assert.NoError(t, err)
+	require.NotEmpty(t, proxyOpts.ProxyPrefix)
+	proxy, err = NewOAuthProxy(proxyOpts, func(s string) bool { return false })
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	proxyTests := []struct {
+		name             string
+		url              string
+		proto            string
+		host             string
+		uri              string
+		expectedRedirect string
+	}{
+		{
+			name:             "proxied request outside of ProxyPrefix redirects to proxiexd URL",
+			url:              "https://oauth.example.com/foo/bar",
+			proto:            "https",
+			host:             "a-service.example.com",
+			uri:              "/foo/bar",
+			expectedRedirect: "https://a-service.example.com/foo/bar",
+		},
+		{
+			name:             "proxied request with query preserves query",
+			url:              "https://oauth.example.com/foo?bar",
+			proto:            "https",
+			host:             "a-service.example.com",
+			uri:              "/foo?bar",
+			expectedRedirect: "https://a-service.example.com/foo?bar",
+		},
+		{
+			name:             "proxied request under ProxyPrefix redirects to root",
+			url:              "https://oauth.example.com" + proxy.ProxyPrefix + "/foo/bar",
+			proto:            "https",
+			host:             "a-service.example.com",
+			uri:              proxy.ProxyPrefix + "/foo/bar",
+			expectedRedirect: "https://a-service.example.com/",
+		},
+		{
+			name:             "proxied request with port under ProxyPrefix redirects to root",
+			url:              "https://oauth.example.com" + proxy.ProxyPrefix + "/foo/bar",
+			proto:            "https",
+			host:             "a-service.example.com:8443",
+			uri:              proxy.ProxyPrefix + "/foo/bar",
+			expectedRedirect: "https://a-service.example.com:8443/",
+		},
+		{
+			name:             "proxied request with missing uri header would still redirect to desired redirect",
+			url:              "https://oauth.example.com/foo?bar",
+			proto:            "https",
+			host:             "a-service.example.com",
+			uri:              "",
+			expectedRedirect: "https://a-service.example.com/foo?bar",
+		},
+	}
+	proxyTestsWithRdQuery := []struct {
```

Comment:
Let's get all these test tables merged into a single one so you can appropriately test intermixing headers & querystring settings so we have strong assurances of order of operations behavior of it controlled in our tests.

I think you can add a `rd string` entry to the test table and use either `""` as the case for no `rd` querystring or some magic const in case you want to test `""` resulting in `?rd=` in the path.

```
if tt.rd != "" {
    req.Form.Add("rd", tt.rd)
}
```